### DISCUSSION TITLE
Is there an EASY way  to process / display (on WWW) data from ADS1115 sensor?

### DISCUSSION THREAD

**Discussion by laurentppol** [arendst/Tasmota#15664] [Q&A] [answered]
Hi there, 
question is in topic, details here:
- I have 3 current measuring channels (using ACS712-30A sensors). 
- They are Hall sensors with output voltage proportional to current. At 0A voltage is 1/2 Vcc (this is why my 4th ADS channel measures Vcc - 5V).
- ADS is connected to ESP8266 trough voltage level shifter (as ACSes are 5V devices, ADS is 3/5V, ESP is 3.3V).
ADS is in default config (4 single ended channles, 6.144V full scale).

Now in Tasmota WWW I can see RAW values from ADC: 11850 for 0A ACS, 24000 for 5V (readings varies very much - bad Vcc decoupling?).
And I would like to obtain current values (if possible simple way), or at least input voltages in  V (mV).

My (quite complicated but working well for other sensor readings like DSes) method is to put all MQTT into PostgreSQL table, then parse using SQL, then display with Perl + Apache.

**Comment by barbudor** [comment, marked as answer]
If you want to display things on Tasmota web page, you should explore Scripting-Language
Is that easy? I'll you discover it

If you are not specifically looking at having that on Tasmota web page, there are many possibilities
Here are some that comes to mind
* local
  - tasmota to InfluxDb database + grafana
  - tasmota to mqtt to node-red with node-red dashboard
  - home automation software such as Home Assistant, OpenHAB, Iobroker... 
* cloud : there are a few platforms with mqtt input and display capabilities.
  - Adafruit.IO
  - Flespi.io should have such capabilities
 Probably many more

**Reply by laurentppol** [reply]
Which language You mean?
- Berry (haven't looked at yet)
- "script" - cryptic worse than FORTH was...?

[OT]:
For what You named "local" I have described "my" (working) way ;P. 
Quite long, as it passes Tasmota -> Mosquitto -> Perl -> PostgreSQL -> SQL -> Perl -> Apache.

For me cloud is to avoid at any price -> security reasons.

And for OpenHAB, I am learning this, but do You know, it is REALLY HEAVY: You CAN'T start it (version 2!) on an SBC with 1GB RAM! (ex  O/RPi smaller models). It is not responding in (human) realtime. I deployed (OH2) on OrangePi 3 (2GB RAM, 4 core 1.8GHz), works nicely, but has no even concept of security. :(
I plan to use OH3 on OPi4 (4GB RAM, 6 core 2GHz), but have "newbie problems". At least it has some concept of user / admin.

**Reply by laurentppol** [reply]
Berry looks OK.
Why it is only for ESP32?

[personally have spare ESP32-CAM, can use it, currently use WeMos D1 mini Pro]

**Reply by sfromis** [reply]
ESP8266 does not have the capacity (RAM and program space) for Berry + Tasmota.

**Comment by pkkrusty** [comment]
I like @barbudor suggestion of Tasmota->InfluxDB->Grafana. My Raspberry Zero 2 W runs everything easily, and InfluxDB driver for Tasmota makes it easy to set up. Grafana is very flexible and pretty. 

I'm using ADS1115 to measure lithium cells and then do the math internally for charge limiting, and also in grafana to scale correctly (wanted to try both options)

![image](https://user-images.githubusercontent.com/79770016/169681271-1552dd1f-4225-4b2f-8106-7be3788f2dc1.png)

![image](https://user-images.githubusercontent.com/79770016/169681282-0e9f82e1-fb38-4677-88bf-9c1b48709cce.png)

**Reply by Chris-The-Tuner** [reply]
I would love how you get the Voltage calculated from the raw (full_scale) values Tasmota is giving via MQTT.
I try my best since days but i'm unable to get a math operation that does give me what i need :)

**Reply by pkkrusty** [reply]
I'm doing it using rules. In my specific case, each cell is connected with a voltage divider to the ADS1115, thus each reading has a unique equation to map raw values to cell voltage. I take four or more readings at various states of charge and then do a linear regression analysis to get a formula.

As an example, I took readings for cell 1: 

5418 | 3.922 ; 5373 | 3.89 ; 5356 | 3.88 ; 5242 | 3.80 ; 5014 | 3.65

And then used Excel (or https://www.omnicalculator.com/statistics/linear-regression) to get a nice formula. 

Not sure how well Tasmota deals with floats though, since calculations seem off sometimes. If I use the resulting formula of RawValue*0.00067117+0.28544, values aren't very accurate. If however, I change the formula around so it's dealing with larger numbers i.e. RawValue/1489.95+0.28544, it's better. Statistical significance is quite high (F=0.0000031), so I'm confident that the relationship is linear for the range I need. 

I may need to do some reading into float calculations, and maybe multiply everything by 1000 for the calcs and then divide by 1000 at the end.

My rulesets are as follows for my system:

Rule1 
on system#init do backlog power3 1;ruletimer3 4; ruletimer4 5; sensor12 s1 endon 
On ina226-1#Voltage!=%var1% do var1 %value% endon 
On ina226-1#Current!=%var2% do var2 %value% endon 
On rules#timer=3 do backlog displaytextnc %var1%; ruletimer2 2 endon 
On rules#timer=2 do backlog displaytextnc %var2%; ruletimer3 2 endon 
On tele-ads1115-48#a0 do var3 %value% endon 
On tele-ads1115-48#a1 do var4 %value% endon 
On tele-ads1115-48#a2 do var5 %value% endon 
On tele-ads1115-49#a0 do var6 %value% endon 
On tele-ads1115-49#a1 do var7 %value% endon 
On tele-ads1115-49#a2 do var8 %value% endon

Rule2
On var16#state<3.3 do power2 0 endon
On var15#state>4 do power1 0 endon
On var15#state<3.93 do power1 1 endon
On var16#state>3.5 do power2 1 endon
On var15#state>4.1 do backlog power1 0; publish 24v-charger-2/stat/alert {"Voltage":"%value%","Current":"%Var2%"}; ruletimer1 10 endon
On ina226-1#current<-8 do backlog power2 0; publish 24v-charger-2/stat/alert {"Error":"Discharge current over 8a"} endon
On rules#timer=1 do publish 24v-charger-2/stat/alert {"Voltage":"%Var1%","Current":"%Var2%"} endon 
On time#minute=360 do power3 1 endon
On time#minute=1320 do power3 0 endon
on ds18b20#Temperature>36 do power1 0 endon
on ds18b20#Temperature<34 do if (%var15%<4) power1 1 endif endon

Rule3
on rules#timer=4 do backlog var9=var3/1489.95+0.2854; var10=(var4/1558.97+0.8734)-(var3/1489.95+0.2854); var11=(var5/1366.83+0.1412)-(var4/1558.97+0.8734); var12=(var6/1537.72+1.9751)-(var5/1366.83+0.1412); var13=(var7/1479.25+2.5554)-(var6/1537.72+1.9751); var14=(var8/1458.31+2.3377)-(var7/1479.25+2.5554); event t1 endon
on event#t1 do backlog publish 24v-charger-2/tele/voltages {"Cell-1":%var9%,"Cell-2":%var10%,"Cell-3":%var11%,"Cell-4":%var12%,"Cell-5":%var13%,"Cell-6":%var14%,"Max":%var15%,"Min":%var16%}; ruletimer4 10 endon
On var9#state do var15 %value% endon
On var10#state>%var15% do var15 %value% endon
On var11#state>%var15% do var15 %value% endon
On var12#state>%var15% do var15 %value% endon
On var13#state>%var15% do var15 %value% endon
On var14#state>%var15% do var15 %value% endon
on var9#state do var16 %value% endon
On var10#state<%var16% do var16 %value% endon
On var11#state<%var16% do var16 %value% endon
On var12#state<%var16% do var16 %value% endon
On var13#state<%var16% do var16 %value% endon
On var14#state<%var16% do var16 %value% endon

**Comment by gemu2015** [comment]
check this
https://github.com/arendst/Tasmota/discussions/15767
